Support work for issue #12 - required changes to how plugins work

GnaspGames opened the pull request:
Support work for issue #12 - required changes to how plugins work so that additional data can be sent to plugins in the future. 

Example of a plugin now:

```
var Test = {};

Test.Install = function(smelt)
{
    smelt.addSupportModule("bang-commands-setup.mcc");
}

Test.Execute = function(smelt)
{
     // Arguments available on smelt.args
    var input = smelt.args.join(" ");

    // Add new command blocks using smelt.addCommandBlock()
    smelt.addCommandBlock("say Hello World", { type:"impulse", auto:true, conditional:false });
    smelt.addCommandBlock("say " + input, { type:"chain", auto:true, conditional:false });

    // Read variables previously set using smelt.getVariable()
    var testInputVar = smelt.getVariable("$testInputVar");
    if(testInputVar)
    {
        smelt.addCommandBlock("say testInputVar = " + testInputVar);
    }

    // Set variables to be used later using smelt.setVariable()
    smelt.setVariable("$testOutputVar", input);

     // Read the current smelt config using smelt.settings
    // For example: smelt.settings.Output.ShowDebugInfo shows that the user
    // wants additional debug info.
    if(smelt.settings.Output.ShowDebugInfo)
    {
        console.log("Setting $testOutputVar to:" + input);
    }
}
module.exports = Test; 
```

Each plugin should implement an `Execute()` function and an `Install()` function. `Execute()` takes the place of the old approach. 

By passing a single `smelt` object into these methods, it means that new properties and methods can be added more easily in the future. Affording backwards compatibility for the existing approach. 

The new `Install()` function will be used to register any support modules that the map maker should install into their world. By separating this out, we can add a call to this function by itself in other situations, such as when the plugin is installed, or if the command is implement to just give us the support module for a plugin. 

Building on top of this; I intend to add properties to the `smelt` object passed to plugins so that contextual data is available, such as what the previous command block was, the coordinates relative to the corner of the module, or relative to the next command block.

GnaspGames committed:
Change bangcommands to use "smeltObj"

GnaspGames committed:
Merge branch 'dev' into feature-12

GnaspGames committed:
Merge branch 'dev' into feature-12

GnaspGames committed:
Merge branch 'dev' into feature-12

GnaspGames committed:
Merge branch 'dev' into feature-12

GnaspGames committed:
Merge branch 'dev' into feature-12

GnaspGames committed:
Remove comments in CheckforVariables

GnaspGames committed:
Merge branch 'master' into feature-12

GnaspGames committed:
Update plugins; new interface to implement

Each plugin should implement two methods:

Install() and Execute()

Install() will be used to define support modules to be installed into the map makers world.

Execute() will be what's run when the plugin command it called.

GnaspGames committed:
Update demo plugin and add missing semi-colons

mrjvs commented:
I Did not know how plugins worked before. But this way you can easily transfer data from smelt to the plugin. It's looks easy to work with and responsive (if that is the right term).

GnaspGames committed:
Update !math correctly

GnaspGames committed:
Keep !scoreTp using old approach for testing

GnaspGames committed:
Changes to smelt paramter sent to bangs/plugins...

* smelt.getVariable() added.
* smelt.addVariable() renamed smelt.setVariable() (get/set not get/add)
* smelt.settings allows access to current config settings:

e.g. use `smelt.settings.Output.ShowDebugInfo` to test whether additional debug info should be written to the console.

GnaspGames commented:
smelt object now has: `getVariable()`, `setVariable()` and `settings` property. See edited first post.

GnaspGames committed:
Merge branch 'dev' into feature-12

M4GNV5 committed:
fix plugin placed commands in xx-chain mode not changing type to chain

M4GNV5 committed:
add smelt.addInitCommand for plugins to run commands when installing the module and add pre.js that runs the args as a initCommand

M4GNV5 committed:
use initCommands with the math plugin for adding the math objective and setting const values

GnaspGames committed:
Merge pull request #50 from M4GNV5/feature-12

Add smelt.addInitCommand for plugins

M4GNV5 committed:
allow using variables withou {} but with a word border.

GnaspGames committed:
Merge pull request #51 from M4GNV5/feature-12

allow using variables without {} but with a word border.

GnaspGames committed:
Add build.sh for me to make sure I use LF ending
